core/src/main/java/hudson/model/AbstractItem.java

```diff
@@ -511,14 +514,48 @@ public boolean hasPermission(Permission p) {
      */
     public synchronized void save() throws IOException {
         if(BulkChange.contains(this))   return;
-        getConfigFile().write(this);
+        synchronized (saving) {
+            saving.add(this);
+        }
+        try {
+            getConfigFile().write(this);
+        } finally {
+            synchronized (saving) {
+                saving.remove(this);
+            }
+        }
         SaveableListener.fireOnChange(this, getConfigFile());
     }
 
     public final XmlFile getConfigFile() {
         return Items.getConfigFile(this);
     }
 
+    private static final Set<AbstractItem> saving = new HashSet<>();
+
+    private Object writeReplace() {
+        synchronized (saving) {
+            if (saving.contains(this)) {
+                return this;
+            } else {
+                LOGGER.log(Level.WARNING, "JENKINS-45892: reference to {0} being saved but not at top level", this);
+                return new Replacer(this);
+            }
+        }
+    }
+
+    /** Not {@link Serializable} for now, since we are only expecting to use this from XStream. */
+    private static class Replacer {
+        private final String fullName;
+        Replacer(AbstractItem i) {
+            fullName = i.getFullName();
+        }
+        private Object readResolve() {
+            // May or may not work:
+            return Jenkins.getInstance().getItemByFullName(fullName);
```

Comment:
From what I see this call may happen when Jenkins instance is null and when the jobs are not fully loaded. It would be better to handle this cases explicitly 🐛